---
title: "Rpack"
subtitle: "Experience around the course"
author: "David Nathan Ohayon"
date: "2023-12-04"
output: 
  ioslides_presentation:
    #incremental: true
    widescreen: true
    fullscreen: true
    smaller: true
    logo: ../data/UNIGE.png
    css: styles.css
    revealjs::revealjs_presentation:
      theme: night
      center: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r, echo = FALSE, message=FALSE}
library(ProteomicsTools)
```

## Few words about me

- Biologist passionated about (bio)-informatics
  - Started a Master in Bioinformatics (2 years ago)
  - Bioinformatics: A very wide field
  - Not (only) about programming
- Being self-taught in programming is hard:
  - Gives you bad habits
  - Not the best way to learn 
  - Bioinformatics in Geneva lacks some programming courses

## Few words about my project  {.flexbox .vcenter}

- **Proteomic experiment and analysis** 
  - **Purpose**: Compare 5 methods of protein extraction 
- **Question**:
  - Is there a bias between one given method and one given protein ?
- 5 methods of protein extraction in order to compare those methods:
  - RIPA
  - NucleoSpin
  - Tris.HCl
  - Triton.X100
  - Urea
- The same cell has been lysed with those methods (a human melanoma cell line [G-361])
- Protein abundance have been estimated using MS technology
- **Particularity**: Not gene expression but protein abundances 

---


### Input Table (log2 protein abundances)

```{r}
library(DT)
rownames(log2) <- log2[,1]
datatable(log2[,-1], options = list(scrollX = TRUE, scrollY = "350px"))
```


## First part of the Master (2021-2022)

- **Idea**: Is there a bias given the subcellular location of the proteins?
- Datamining of the NeXtProt Database 
- As being a self-taught programmer:
  - used to make python scripts 
  - was happy enough that my scripts work
  - had to wait a lot for my script to be over, since the use of an API and for-loops

---

### Exemple with 2 for-loops, that used to take a lot of time

```{python, eval=FALSE, echo=TRUE}
def processglyco(proteins, filename):
    with open(filename, 'w') as output:
        writer = csv.writer(output, delimiter=";")
        writer.writerow(['ProteinID', 'GlycoSites'])
        for protein in proteins:
            response = requests.get(f"https://api.nextprot.org/entry/NX_{protein}/glycosylation-site.json")
            json = response.json()
            glycosites = []
            try:
                glycosite = json["entry"]["annotationsByCategory"]["glycosylation-site"]
                if glycosite != {}:
                    for elements in glycosite:
                        if elements['qualityQualifier'] == 'GOLD' or 'SILVER':
                            glycotype = elements["cvTermName"]
                            if glycotype not in glycosites:
                                glycosites.append(glycotype)
                    print(protein, glycosites)
                    writer.writerow([protein, glycosites])
            except KeyError:
                pass

```

## Second part of my master (2022-2023)

- **Statistics and Probability (Autumn 2022)**:
  - have shown me new way of analyzing data in R (PCA...)
- **Statistics and Probability (Spring 2023)**
  - have shown me how to develop in R
  - R is a powerful tool when you know how it works
  - created my own package of analyzing and visualizating data
  - helped me a lot to find results for my thesis 
- **Benefits**: 
  - Private meetings with custom made help 
  - Workload to succeed the course is ridiculous (for people that have a dataset to analyze)
  - Indeed, I was working for the course and my master at the same time (analyzed dataset from my master)
  - Although it is not the only benefit (people that didn't have a dataset have learnt as much as me)
- Examples of my package:
  
---
  
### Pairwise t-test for each pair of method

```{r}
datatable(ttest, options = list(scrollX = TRUE, scrollY = "350px"))
```

---

### Pairwise log2-foldchange for each pair of method

```{r}
datatable(foldc, options = list(scrollX = TRUE, scrollY = "350px"))
```

---

### VolcanoPlots for each method, with extra highlight 
<div style="height: 470px; overflow: scroll;">
```{r, echo = TRUE, highlight=FALSE, warning=FALSE, message=FALSE, fig.width=40, fig.height=5, dpi=72}
volcano_all(ttest, foldc, datamined, "TMregions", path = "../examples/")

```
</div>

---

### Enrichment analysis of my pairwise t-tests 

```{r}
datatable(ftest, options = list(scrollX = TRUE, scrollY = "350px"))
```

---


### DotPlot to visualize enrichment analysis

```{r, echo = TRUE, highlight=FALSE, fig.width=10, fig.height=5.2}
fisher_dotplot(ftest, path = "../examples/")
```

---

### This is not over: Optimizing my scripts and traducing my python scripts in R

```{r, echo=TRUE, eval=FALSE}
get_nextprot_subcell <- function(protname) {
  api_endpoint <- "https://api.nextprot.org/entry/"
  api_end <- "/subcellular-location.json"
  protein_id <- paste0("NX_", protname)
  response <- httr::GET(paste0(api_endpoint, protein_id, api_end))
  if (response$status_code == 200) {
    json_data <- httr::content(response, as = "text")
    parsed_data <- (jsonlite::fromJSON(json_data))
    data2mine <- parsed_data$entry$annotationsByCategory$`subcellular-location`
    data2mine <- data2mine[data2mine$qualityQualifier == "GOLD",]
    data2mine$cvTermName
  } else {return(paste0("Error: ", response$status_code))}
}

datamining_assembling <- function(df, colname, fun) {
  data2assemble <- df
  data2assemble$column <- pbapply::pblapply(data2assemble$UniProtID, function(prot) fun(prot))
  names(data2assemble)[names(data2assemble) == "column"] <- colname
  return(data2assemble)
}
```

---

### Output of my datamining function in R

```{r}
datatable(fisher_subcell, options = list(scrollX = TRUE, scrollY = "350px"))
```

---


```{r, echo = FALSE, highlight=FALSE, fig.width=10, fig.height=5.2}
fisher_dotplot(fisher_subcell, path = "../examples/")
```


## Quick conclusion

This package has turned out really helpful to:

- Understand and advance my studies
- Give me new ideas about my project 
- Made myself an ID card with this package

Around R:

- Has shown me new possibilities of R
- Has shown me how good RStudio is (Git, package test...)
- Learnt new programmatic approaches (apply, melting table...)
